app/http.php

```diff
@@ -0,0 +1,366 @@
+
+        // If no health check, mark all as online
+        $executors = \explode(',', (string) App::getEnv('OPEN_RUNTIMES_PROXY_EXECUTORS', ''));
+
+        foreach ($executors as $executor) {
+            $executorStates->set($executor, [
+                'status' => 'online',
+                'hostname' => $executor,
+                'state' =>  \json_encode([])
+            ]);
+        }
+    }
+});
+
+Console::log('State of executors at startup:');
+
+go(function () use ($register) {
+    $executorStates = $register->get('executorStates');
+
+    $executors = \explode(',', (string) App::getEnv('OPEN_RUNTIMES_PROXY_EXECUTORS', ''));
+
+    foreach ($executors as $executor) {
+        $state = $executorStates->exists($executor) ? $executorStates->get($executor) : null;
+
+        if ($state === null) {
+            Console::warning('Executor ' . $executor . ' has unknown state.');
+        } else {
+            Console::log('Executor ' . $executor . ' is ' . ($state['status'] ?? 'unknown') . '.');
+        }
+    }
+});
+
+Swoole\Event::wait();
+
+// Wildcard action
+App::error()
+    ->inject('request')
+    ->inject('response')
+    ->inject('roundRobinAtomic')
+    ->inject('executorStates')
+    ->action(function (Request $request, Response $response, Atomic $roundRobinAtomic, Table $executorStates) {
+        $secretKey = \explode(' ', $request->getHeader('authorization', ''))[1] ?? '';
+
+        if (empty($secretKey)) {
+            throw new Exception('Incorrect proxy key.', 401);
+        }
+        if ($secretKey !== App::getEnv('OPEN_RUNTIMES_PROXY_SECRET', '')) {
+            throw new Exception('Incorrect proxy key.', 401);
+        }
+
+        $roundRobinIndex = $roundRobinAtomic->get();
+
+        $algoType = App::getEnv('OPEN_RUNTIMES_PROXY_ALGORITHM', '');
+        $algo = match ($algoType) {
+            'round-robin' => new RoundRobin($roundRobinIndex - 1), // Atomic indexes from 1. Balancing library indexes from 0. That's why -1
+            'random' => new Random(),
+            default => new Random()
+        };
+
+        $balancing = new Balancing($algo);
+
+        $balancing->addFilter(fn ($option) => $option->getState('status', 'offline') === 'online');
+
+        foreach ($executorStates as $state) {
+            $balancing->addOption(new Option($state));
+        }
+
+        $body = \json_decode($request->getRawPayload() ? $request->getRawPayload() : '{}', true);
+        $runtimeId = $body['runtimeId'] ?? null;
+        // TODO: @Meldiron Use RuntimeID in CPU-based adapter
+
+        $option = $balancing->run();
+
+        if ($option === null) {
+            throw new Exception('No online executor found', 404);
+        }
+
+        if ($algo instanceof RoundRobin) {
+            $roundRobinAtomic->cmpset($roundRobinIndex, $algo->getIndex() + 1); // +1 because of -1 earlier above
+        }
+
+        $executor = [];
+
+        foreach ($option->getStateKeys() as $key) {
```

Comment:
It's better to stick to primitive types everywhere we can. It simpler and more predictable. You're only using this block to access $executor['hostname'] and it doesn't seem to provide much help. If we need to convert an object to assoc array there are better patterns we can discuss. For this case, I believe using $option->getState('key', 'default') could work better.

Reply:
Problem with `getState` is that it requires a key. The idea here is to get a state array. All of it.

Should I remove `getStateKeys` method and instead add `getStates` that would simply return the whole array?

Reply:
Suggestion showed in this PR: https://github.com/utopia-php/balancing/pull/4

Reply:
Yeh, I think that would make the signature cleaner for now.

Reply:
Implemented without getStateKeys